@xen-orchestra/lite/src/libs/story/story-param.ts

```diff
@@ -0,0 +1,436 @@
+import {
+  faFloppyDisk,
+  faRocket,
+  faShip,
+  faTrash,
+} from "@fortawesome/free-solid-svg-icons";
+import mixin, { type MixinAbstractConstructor } from "@/libs/mixin";
+import type { Widget } from "@/libs/story/story-widget";
+import {
+  boolean,
+  choice,
+  number,
+  object,
+  text,
+} from "@/libs/story/story-widget";
+
+function WithType(Base: MixinAbstractConstructor<BaseParam>) {
+  abstract class WithType extends Base /*implements HasType*/ {
+    #type: string | undefined;
+
+    type(type: string) {
+      this.#type = type;
+
+      return this;
+    }
+
+    abstract getTypeLabel(): string;
+
+    getType() {
+      return this.#type;
+    }
+  }
+
+  return WithType;
+}
+
+function WithWidget(Base: MixinAbstractConstructor<BaseParam>) {
+  abstract class WithWidget extends Base /*implements HasWidget*/ {
+    #widget: Widget | undefined;
+
+    abstract guessWidget(): Widget | undefined;
+
+    widget(widget?: Widget): this {
+      this.#widget = widget === undefined ? this.guessWidget() : widget;
+
+      return this;
+    }
+
+    getWidget() {
+      return this.#widget;
+    }
+
+    hasWidget() {
+      return this.#widget !== undefined;
+    }
+  }
+
+  return WithWidget;
+}
+
+abstract class BaseParam {
+  #help: string | undefined;
+  readonly #name: string;
+  #presetValue: any;
+
+  protected abstract getNamePrefix(): string;
+
+  protected abstract getNameSuffix(): string;
+
+  constructor(name: string) {
+    this.#name = name;
+  }
+
+  get name() {
+    return this.#name;
+  }
```

Comment:
Why not making it public?

Reply:
Because in general, I prefer having a private property then control its read/write access when needed.